### ISSUE TITLE
Can someone help me pls: Error:spawn jdb ENOENT

### ISSUE THREAD

**Issue by DanJ210** [DonJayamanne/javaVSCode#19]
I get the error above and I am somewhat new to what I'm supposed to set values to but this extension looked very awesome and simple. I took a screenshot of the launch.json that was shown and though mine looks different, the values are default just as in the example. So am I missing something? I took screenshots of exactly what I'm trying to do and run. If anyone would be kind enough to help me figure what I am missing, that would be great, thank you in advance.

https://1drv.ms/i/s!AlnGeQSYLUVaoo1djC8YfdYT7diUpQ
https://1drv.ms/i/s!AlnGeQSYLUVaoo1eki3dEciFuNcYwA

**Comment by faustinoaq** [comment]
Hi @DanJ210 :raising_hand_man: 

I didn't see `jdkPath` in your `launch.json`, maybe `jdb.exe` is already in your path.

You have Scanner class in your code for input, then external console is needed

Try this:

``` json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Java",
      "type": "java",
      "request": "launch",
      "stopOnEntry": true,
      "externalConsole": true,
      "jdkPath": "C:/Program Files/Java/jdk1.8.0_112/bin",
      "cwd": "${workspaceRoot}",
      "startupClass": "TestingJava",
      "options": [
        "-classpath",
        "${workspaceRoot}"
      ]
    }
  ]
}
```

The project structure:

```
MyProjectJava/
    TestingJava.java
    TestingJava.class
    .vscode/
        launch.json
```

**Comment by DanJ210** [comment]
@faustinoaq Hey Thanks! You got me further than ever before! I'm getting one thing though, debugger failed to attatch: handshake failed - connection prematurely closed.

https://1drv.ms/i/s!AlnGeQSYLUVaoo1hRNDkk-RGwcFyPA

I did copy the launch.json code you gave me but left the second part, then got rid of the second part to see if that was it. I also had to change my JDK to 111, I didn't download 112 because I wasn't sure if I should be using that or not. You showed me what I needed to link the jdk path, thank you!!!

I'm going to install 112 and see if that helps at all. Do I need to add anything to the TestingJava.class file? Leave it blank? I may even create a file structure just like you have and see if that's the reason. Any further help is much appreciated and thanks so much already for the assistance. Oh and I did remove the Scanner usage for now, all I've got is a for loop.

**Comment by DanJ210** [comment]
Well I did remove my old java installations and installed the 112, I then created the folder structure exactly like you have it but I'm still getting the error above. Sorry if I'm being noobish and I don't want to bug but any ideas would make my day! Thanks @faustinoaq for the help already.

**Comment by faustinoaq** [comment]
@DanJ210 Hi :raising_hand_man: 

I can see the error too :point_right: `Debugger failed to attach`, probably because `cmd.exe` is executed too early. cc @DonJayamanne :eyes: 

But The error don't break the debuger.

![java-debugger](http://i.imgur.com/TDn081X.png)

I like your color theme :+1: Can you share the theme name?

**Comment by DanJ210** [comment]
![vscode - theme](https://cloud.githubusercontent.com/assets/16159685/20099861/b43d904e-a588-11e6-8905-05122edaad3a.png)

Sure! There's the theme. I found it by accident but glad I did because the colors are just gorgeous. There's VScode - icons also which gives you nice file and folder icons, that's a great one too.

Now onto your screenshot, how did you continue like that because I can't, I can't choose next or continue or nothing. Mine just stops. I even removed the scanner stuff and one more item that I highlighted in my screenshot that I see you didn't have but still no change. Do I need settings.json? tasks.json? Any ideas what I'm failing to see here? Thanks again.

![vscode - java](https://cloud.githubusercontent.com/assets/16159685/20099947/02a56536-a589-11e6-9047-764af673162b.png)

**Comment by faustinoaq** [comment]
Hi @DanJ210 Thanks for the theme!

You don't need `tasks.json` or `settings.json`, just `launch.json` is needed by Java Debuger and _obviously_ the Java compiled bytecode (.class file) with debuger flag `-g`.

Try with `externalConsole` and `Scanner` disabled (commented)

![internal-debug-console](http://i.imgur.com/OxNzMG6.png)

_Optional:_ You can use `tasks.json` to compile the Java sources:

``` json
{
  "version": "0.1.0",
  "command": "javac",
  "isShellCommand": true,
  "args": [
    "-g",
    "-cp",
    "${workspaceRoot}",
    "TestingJava.java"
  ],
  "showOutput": "always"
}
```

**Comment by DanJ210** [comment]
@faustinoaq Maybe what I am missing is the debugger flag -g because I am so new to this and know nothing about that. Actually, I commented out scanner and commented out external console which turns off the external console window I just learned, and I get all this in the debug window. Maybe that helps understand what my error is? From what it sounds like, I'm pretty sure that I'm missing the debugger flag or something in my TestingJava.class file which is completely blank.

I'm sorry to keep bugging you with this but I'm just so new and lost! Lol. It's driving me nuts trying to figure out what I'm doing wrong! I'm pretty sure it's that bytcode java compiler debug flag thing you mentioned.

![vscode- javatest1](https://cloud.githubusercontent.com/assets/16159685/20116834/dddb0718-a5cb-11e6-8d7d-b2e7dbe65325.png)

**Comment by faustinoaq** [comment]
@DanJ210 Don't worry I'm learning too 😄 

The error `java.lang.ClassFormatError: Truncated class file`  is because bad bytecode compilation.
Seems that your .class file is empty

Open `cmd.exe` in the workspace directory and execute
`"C:/Program Files/Java/jdk1.8.0_112/bin/javac.exe" -g TestingJava.java`

I'm using `vscode` for debugging Java (This Java Debuger) and C++ (with [Native Debug](https://marketplace.visualstudio.com/items?itemName=webfreak.debug) for .exe) because I like the vscode design. Also vscode is the only editor where I can edit many programming languages at once with editor/ide features out the box (or minimum configuration).

By the way, I did my own extension for Java linting. When you want review it 👉  [Java Linter](https://marketplace.visualstudio.com/items?itemName=faustinoaq.javac-linter) (basic error checking per file)

**Comment by DanJ210** [comment]
@faustinoaq Awesome! And I've actually seen the Java Linter extension and I will try it. Figured once I get this running I can use it! I'll install it now.

Ok so first time I've done the above but I actually did get it. It looks like that did fill out my TestingJava.class file with binary stuff so this looks great! But now I get this error about JRE stuff, which I think I may have to install JRE separate from JDK. They are two different things from what I understand. So I will try installing JRE and see if that changed anything, do you see the same issue?
![vscode - java 4](https://cloud.githubusercontent.com/assets/16159685/20120328/6b28b6c0-a5da-11e6-9ba9-83128022af1a.png)

I am using VScode to learn in class and practice on my own. I like how lightweight it is and I like how the extensions can change/add functions and features. I really like the colors. I'm also having to do C#, JS, and HTML/CSS so it's nice using it for that as well. Plus the intellisense helps a lot.

**Comment by DanJ210** [comment]
Whoops, I think I left the jdk path commented out which is why I got the error! I'm leaving my post above just for future help if anyone needs! But the error above is because I had my JDK path still commented out for testing something, my fault!

BUT OMG IT WORKS!
![vscode - java 5](https://cloud.githubusercontent.com/assets/16159685/20120726/f62f50de-a5db-11e6-8293-106c34603277.png)

You don't know how thankful I am dude, I really really appreciate all your help. I'm installing Java Linter now via your link to make sure I get the right one and I will review it. Would be cool to keep in touch! Once MSFT teams comes out, we can have a group of programmers like a chat room on it. There's also discord and slack which I use and is great for keeping in touch with other programmers. But this is so awesome thank you so much, I just can't say it enough!

**Comment by DanJ210** [comment]
@faustinoaq Is there any kind of instruction list for beginners about Linting? I suppose that's a silly question. I'm going to look it up and review it now but I do see that in your example video/gif thing, you are highlighting and changing multiple lines at the same time. What do I need to know?

I also see your settings down below. I need to open workspace settings to create a settings.json file and then copy those into the settings? I have done that and so far no errors but I am also unsure what exactly I did! So I am going to spend this night learning and researching all of the VScode tutorials on what I'm doing. Thanks again.

Oh and I guess I should close this issue.

**Comment by faustinoaq** [comment]
> Would be cool to keep in touch! Once MSFT teams comes out, we can have a group of programmers like a chat room on it. There's also discord and slack which I use and is great for keeping in touch with other programmers.

@DanJ210 Great 👍 My email 👉 `faustinoaq at outlook dot com`

**Comment by faustinoaq** [comment]
> I also see your settings down below. I need to open workspace settings to create a settings.json file and then copy those into the settings? I have done that and so far no errors but I am also unsure what exactly I did! So I am going to spend this night learning and researching all of the VScode tutorials on what I'm doing.

@DanJ210 You can open an issue in my repo when you want 👉 [Java Linter issues](https://github.com/faustinoaq/vscode-javac-linter/issues)

> Oh and I guess I should close this issue.

Yes 👍

**Comment by DanJ210** [comment]
@faustinoaq Ok I will do that, one more question before I bother someone else!

But looks like I have to run the binary install every time I add new code which is more than fine. When you debug, can you press the continue button? Cause mine just exits when I do. Otherwise it stops at the first statement and I have to keep clicking step over or into to get it to run. If you get the same thing then I will post another thread.

Otherwise it runs great!

**Comment by faustinoaq** [comment]
> But looks like I have to run the binary install every time I add new code which is more than fine. When you debug, can you press the continue button? Cause mine just exits when I do. Otherwise it stops at the first statement and I have to keep clicking step over or into to get it to run. If you get the same thing then I will post another thread.

@DanJ210 Already I have a related issue 👉 #16

**Comment by customBoks** [comment]
@faustinoaq Hi , i had a problem with them , work's on Kde too and i do , i think the same thing like you but i have always the msg in my debug console "Error: Could not find or load main class TestingJava" , 2day i'am on but nothing. 
![errordebugger](https://cloud.githubusercontent.com/assets/26315783/23779398/90d8b2ec-0540-11e7-9aa1-66169444ae38.png)

![errordebugger2](https://cloud.githubusercontent.com/assets/26315783/23779472/ec4d367a-0540-11e7-827f-5fb80998e39e.png)

**Comment by customBoks** [comment]
....and i don't do nothing with the debugger , start and immediatly i'd this msg , but i don't a have a  problem with compilation and i run my class properly in terminal .

thank's in advance.
Peace

**Comment by DanJ210** [comment]
@customboks You do have the task.json setup properly? I'll have to open the project back up and run through this again. I did save the project so I could use as a reference. I'll find and post all that I have.

**Comment by faustinoaq** [comment]
Hi @DanJ210 
@customBoks already solved his problem here #28 

Greetings :raising_hand_man:
